// frontend_blog_application/src/pages/Signup.jsx
import { useEffect, useState } from "react";
import Base from "../components/Base";
import {
  Button,
  Card,
  CardBody,
  CardHeader,
  Col,
  Container,
  Form,
  FormGroup,
  Input,
  Label,
  Row,
} from "reactstrap";

const Signup = () => {
  const [data, setData] = useState({
    name: "",
    email: "",
    password: "",
    about: "",
  });

  const [error, setError] = useState({
    errors: {},
    isError: false,
  });

  // Additional part to print data on console..
  //   useEffect(() => {
  //     console.log(data);
  //   }, [data]);

  //Handle Change
  const handleChange = (event, property) => {
    //Dynamically setting the values..
    setData({ ...data, [property]: event.target.value });
  };

  // Reset form
  const resetData = () => {
    setData({
      name: "",
      email: "",
      password: "",
      about: "",
    });
  };

  //Submit form
  const submitForm = (event)=>{
	
  };

  return (
    <Base>
      <Container>
        {JSON.stringify(data)}
        <Row className="mt-5">
          <Col sm={{ size: 6, offset: 3 }}>
            <Card color="dark" outline>
              <CardHeader className="text-center">
                <h3>Register</h3>
              </CardHeader>
              <CardBody>
                {/* Creating form */}
                <Form onSubmit={submitForm}>
                  <FormGroup>
                    <Label form="name">Enter name</Label>
                    <Input
                      type="text"
                      id="name"
                      placeholder="Enter name"
                      onChange={(e) => handleChange(e, "name")}
                      value={data.name}
                      required
                    ></Input>
                  </FormGroup>
                  <FormGroup>
                    <Label form="email">Enter email</Label>
                    <Input
                      type="email"
                      id="email"
                      placeholder="Enter email"
                      onChange={(e) => handleChange(e, "email")}
                      value={data.email}
                      required
                    ></Input>
                  </FormGroup>
                  <FormGroup>
                    <Label form="password">Enter password</Label>
                    <Input
                      type="password"
                      id="password"
                      placeholder="Enter password"
                      onChange={(e) => handleChange(e, "password")}
                      value={data.password}
                      required
                    ></Input>
                  </FormGroup>
                  <FormGroup>
                    <Label form="about">About</Label>
                    <Input
                      type="textarea"
                      id="about"
                      placeholder="Write something here.."
                      style={{ height: "150px" }}
                      onChange={(e) => handleChange(e, "about")}
                      value={data.about}
                      required
                    ></Input>
                  </FormGroup>
                  <Container className="text-center">
                    <Button color="primary">Register</Button>
                    <Button
                      onClick={resetData}
                      color="secondary"
                      type="reset"
                      className="ms-2"
                    >
                      Reset
                    </Button>
                  </Container>
                </Form>
              </CardBody>
            </Card>
          </Col>
        </Row>
      </Container>
    </Base>
  );
};

export default Signup;
